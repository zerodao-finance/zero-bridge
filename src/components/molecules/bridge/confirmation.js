import { AiOutlineCheck, AiOutlineClose } from 'react-icons/ai'
import { ethers } from 'ethers'
export const Confirmation = ({transferRequest, back, status}) => {
    transferRequest
    
    return (
        <>
        {transferRequest &&
                        <div className="flex flex-col container h-max bg-white shadow-xl rounded-[30px] justify-center place-items-center gap-3 w-fit pb-4 relative dark:bg-gray-700 dark:text-white">
                            {/* <Confirm /> */}
                            <AiOutlineClose className="absolute top-1 left-1 hover:scale-150 dark:stroke-white" onClick={back}/>
                            <p className=" text-lg font-light text-black  tracking-wider w-full bg-emerald-300 text-center shadow-md rounded-t-md">Confirm Transaction</p>
                            <div className="grid grid-flow-rows grid-cols-2 justify-items-start items-center auto-rows-min min-w-[20rem] max-w-fit mx-10 gap-4">
                                    <p className="text-sm w-fit">asset:</p>
                                    <p className="text-xs w-fit truncate w-2/3 hover:w-full transition-all duration-150">{transferRequest.asset}</p>
                                    <p className="text-sm w-fit">to:</p>
                                    <p className="text-xs w-fit truncate w-2/3 hover:w-full transition-all duration-150">{transferRequest.to}</p>
                                    <p className="text-sm w-fit">underwriter:</p>
                                    <p className="text-xs w-fit truncate w-2/3 hover:w-full transition-all duration-150">{transferRequest.underwriter}</p>
                                    <p className="text-sm w-fit">transfer amount:</p>
                                    <p className="text-xs w-fit truncate w-2/3 ">{ethers.utils.formatUnits(ethers.BigNumber.from(transferRequest.amount), 8)}</p>
                            </div>
                            <p className="capitalize font-thin">fees</p>
                            <div className="grid grid-flow-rows grid-cols-2 justify-items-start items-center auto-rows-min min-w-[20rem] max-w-fit mx-10 gap-4">
                                    <p className="text-sm w-fit">RenVM Fee:</p>
                                    <p className="text-xs w-fit truncate w-2/3 text-red-400">{`-${(0.001 + (ethers.utils.formatUnits(ethers.BigNumber.from(transferRequest.amount), 8) * .15)) }`}</p>
                                    <p className="text-sm w-fit">Zero Arbitrum Fee:</p>
                                    <p className="text-xs w-fit truncate w-2/3 text-red-400">{`-${(0.0015 + (ethers.utils.formatUnits(ethers.BigNumber.from(transferRequest.amount), 8) * .3)) }`}</p>
                                    <p className="text-sm w-fit">Curve Fee:</p>
                                    <p className="text-xs w-fit truncate w-2/3 text-red-400">{`-${ethers.utils.formatUnits(ethers.BigNumber.from(transferRequest.amount), 8) * .04 }`}</p>
                            </div>
                            <div className="flex flex-col p-3 gap-2 rounded-md ring-[1px] ring-gray-300"> 
                                <p className="text-sm w-full text-center font-thin underline">BTC deposit address</p>
                                <p className="text-sm font-light text-center w-full truncate w-full transition-all duration-150 cursor-copy select-all">{transferRequest.gatewayAddress ? transferRequest.gatewayAddress : <CgSpinnerTwoAlt className="mx-auto animate-spin font w-[1.3rem] h-[1.3rem] text-emerald-300" />}</p>
                            </div>
                            <div className="w-2/3 ring-orange-500 ring-2 rounded-md self-center text-center text-[13px] text-gray-100 animate-scale-in-hor-center">
                                <p className="text-orange-500 h-fit">{status ? status : "REMINDER !" }</p>
                                <p className="text-black dark:text-white">Deposit the exact amount of BTC to the Deposit Address</p>
                                
                            </div>
                        </div>
        }
        </>
    )
}