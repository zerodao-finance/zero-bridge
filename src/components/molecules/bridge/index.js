export * from './mechanism'
export * from './confirmation'