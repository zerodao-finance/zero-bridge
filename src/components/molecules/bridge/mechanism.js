import { useState } from "react"
import { TransferTool } from './transfer'
import { ReleaseTool } from './release'



const buttonConfig = {
    "clicked" : "bg-white dark:text-white text-black dark:bg-gray-700 h-full p-3 rounded-t-[20px]",
    "disabled" : "bg-gray-200 dark:bg-gray-800 h-full p-3 text-gray-500 cursor-pointer hover:text-white transition-all delay-150",
}
export const Mechanism = ({ _transactionSender, _context }) => {
    const [tool, toggle] = useState("transfer")
    const toggleTool = (_tool) => {
        toggle(_tool)
    }
    const [ isLoading, sign ] = _transactionSender()

    return (
                <div className='flex flex-col container h-fit bg-white shadow-xl rounded-[30px] justify-center place-items-center gap-3 first:gap-0 w-fit pb-4 dark:bg-gray-700 text-white '>
                    {  
                    isLoading &&
                    <CgSpinnerTwoAlt className="fixed animate-spin w-[3rem] h-[3rem] text-emerald-300" />
                    }
                    <p className="text-lg font-light text-black tracking-wider w-full bg-emerald-300 text-center shadow-md rounded-t-md">Bridge Funds</p>
                        { 
                        <div className={`h-full w-full grid grid-cols-2 grid-flow-rows mb-8 bg-gray-200 dark:bg-gray-800 pt-3 align-center font-light tracking-wider text-sm text-center`}>
                            
                            <div className={tool === "transfer" ? buttonConfig["clicked"] : buttonConfig["disabled"]} onClick={() => toggleTool("transfer")}>
                                Transfer
                            </div>
                            <div className={tool === "release" ? buttonConfig["clicked"] : buttonConfig["disabled"]} onClick={() => toggleTool("release")}>
                                Release
                            </div>
                            
                        </div>
                        }
                        {
                            tool === "transfer" && <TransferTool _isLoading={isLoading} _action={sign} _context={_context}/>
                        }
                        {
                            tool === "release" && <ReleaseTool _isLoading={isLoading} _context={_context} />
                        }
                </div>
                
    )
}