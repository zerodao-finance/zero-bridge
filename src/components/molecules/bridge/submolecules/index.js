export * from './input'
export * from './ratio'
export * from './results'