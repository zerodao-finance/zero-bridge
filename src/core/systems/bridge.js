import { ethers } from 'ethers'
import _ from 'lodash'
import { TrivialUnderwriterTransferRequest, TransferRequest } from 'zero-protocol/dist/lib/zero'
import { updateTransferRequest } from './refresh'
import { Monitor, Observer } from '../tools'
import tools from '../../utils/_utils'
import { getTableHeadUtilityClass } from '@mui/material'
import TransactionCard from '../../components/molecules/TransactionCard'

export class BridgeMonitor extends Monitor {
    /**
     * @attach : (_observer):Observer[]
     * @detach : (_observer): Observer[]
     * @notify : (_group): String
     * 
     */
    constructor(_type){
        super(_type)
    }

    async _create(_to, _value, _ratio){
        this.error = null
        console.log(`\n ${this.type} Subject: creating a Transfer Request `)
        try {
            _to = await _to.getAddress()
        } catch ( error ) {
            this.error = new Error("Please Connect Wallet")
            return this.notify("error_group")
        }

        try {
            ethers.utils.parseUnits(_value, 8)
        } catch {
            this.error = new Error("Invalid Input Value")
            return this.notify("error_group")
        }
        const data = ethers.utils.defaultAbiCoder.encode(
            ["uint256"],
            [ethers.utils.parseEther(parseFloat(String(Number(_value) / 100 * _ratio)).toFixed(8))]
        )

        const asset = tools.asset
        const transferRequest = new TransferRequest({
            to: _to,
            contractAddress: tools.controller.address,
            underwriter: tools.trivialUnderwriter,
            module: tools.zeroModule,
            asset,
            amount: ethers.utils.parseUnits(_value, 8),
            data: String(data)
        })
        this.transferRequest = transferRequest
        return transferRequest
    }

    async load(_transferRequest){
        this.transferRequest = new TransferRequest(_.omit(_transferRequest, ['date']))
        return this._gatewayAddress = await this.transferRequest.toGatewayAddress()
        this.notify("refresh_group")
    }

    async pollTX(_transferRequest){
        this.transferRequest = new TransferRequest(_.omit(_transferRequest, ['date', 'status']))
        this._gatewayAddress = await this.transferRequest.toGatewayAddress()
        this._mint = await this.transferRequest.submitToRenVM()
        this.notify("card_group")

    }

    async sign(signer){
        if (this.error) return
        if (!this.transferRequest || !signer) return this.error = new Error(`${this.type} Subject: ERROR signing: no transfer request or signer`)
        try {
            return this.signature = await this.transferRequest.sign(signer)
        } catch(error){
            console.log("error signing transaction, no wallet connected", error)
        }
    }

    async dry(signer){
        if (this.error) return
        try {
            this._dry = await new TrivialUnderwriterTransferRequest(this.transferRequest).dry(signer.provider, { from : '0x12fBc372dc2f433392CC6caB29CFBcD5082EF494'})
        } catch (error) {
            this.error = new Error("Loan will fail, double check input values")
            return this.notify("error_group")
        }
        this._key = await tools.storage.set({...this.transferRequest, date: Date.now(), dry: this._dry})
        this._event = "transaction_signed"
    }
    
    async gatewayAddress(mock){
        if (this.error) return
        if (mock) return this.gatewayAddress = this._mint.gatewayAddress
        this._gatewayAddress = await this.transferRequest.toGatewayAddress()
        return this.notify("bridge_group")
    }

    async transfer(){
        if ( _.isError(this._dry) ) return this.error = new Error(` ${this.type} Subject: Loan will return Error ...${this._dry}`)
        await this.zeroUser.publishTransferRequest(this.transferRequest)
        updateTransferRequest(this._key, "waiting")
        this._mint = await this.transferRequest.submitToRenVM()
        this.notify("card_group")

    }


    async mockSign(signer){
        console.log(`\n ${this.type} Subject: mocking transfer request signature`)
        if (!this.transferRequest || !signer) return this.error = new Error(`\n ${this.type} Subject: error, no transfer request or signer`)
        return this.notify("bridge_group")
    }


}

/**
 * @group bridge_group
 * @events
 *  - transaction_signed
 *      [ pass signed transferRequest to second conversionTool screen ]
 */
export class BridgeObserver extends Observer {
    constructor(_group){
        super(_group)
        this.nextScreen
        this.prevScreen
    }

    async update (monitor) {
        switch (monitor._event){
            case "transaction_signed":
                let data = monitor.transferRequest
                data.gatewayAddress = monitor._gatewayAddress
                this.nextScreen(monitor.transferRequest)
                break;
        }
    }
}

export class RefreshObserver extends Observer {
    constructor(_group){
        super(_group)
        this.switch
        this.page
    }
    
    async update (monitor) {
        this.switch()
        this.page(monitor.transferRequest)
    }
}


export class TransactionCardObserver extends Observer {
    constructor(_group){
        super(_group)
        this.append
        this.clear

    }
    async update (monitor) {
        monitor._mint.on('deposit', async (deposit) => {
            const hash = deposit.txHash()
            if (deposit.depositDetails.transaction.confirmations >= 6) return updateTransferRequest(monitor._key, "success")
            await this.append(<TransactionCard btc={monitor._gatewayAddress} confs={deposit}/>)
            await deposit
                .confirmed()
                .on("target", target => console.log(`0/${target} confirmations`))
                .on("confirmation", (confs, target) => {
                    console.log(`${confs}/${target} confirmations`)
                    if (Number(confs) === 6){
                        updateTransferRequest(monitor._key, "success")
                    }
                })

        })


        // const deposit = new Promise(async (resolve) => await monitor._mint.on('deposit', resolve)).then(
        //     async (deposit) => 
        //     {
        //         console.log(deposit)
        //         let confirmed = await deposit.confirmed()
        //         console.log(confirmed)
        //         confirmed.on('deposit', async (confs, target) => {
        //             console.log(confs, target)
        //             if (confs == 6){
        //                 console.log(`\nObserver: transaction successful, removing TxnCard from DOM`)
        //                 tools.storage.setStatus(this._key, "success")
        //                 this.clear()
        //             }
        //         })
        //         const signed = deposit.signed();
        //         signed.on("status", async (status) => {
        //             if (status === 'signed') await tools.storage.setStatus(this._key)
        //         })
        //     }
        // )
        // const confirmed = await deposit.confirmed()
    }
}