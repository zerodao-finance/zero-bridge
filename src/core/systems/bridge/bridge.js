import { ethers } from 'ethers'
import _ from 'lodash'
import { TrivialUnderwriterTransferRequest, TransferRequest } from 'zero-protocol/dist/lib/zero'
import { storage } from '../storage'
import { MOCK_TF_RQ } from '../../tools/utilities'
import { EventEmitter } from 'events'
import TransactionCard from '../../../components/molecules/TransactionCard'
import { DataStructures } from '../../tools'
let { Monitor, Observer } = DataStructures

export class BridgeMonitor extends Monitor {
    /**
     * @attach : (_observer):Observer[]
     * @detach : (_observer): Observer[]
     * @notify : (_group): String
     * 
     */

    listener = new EventEmitter()
    constructor(_type){
        super(_type)
    }

    async _create(_to, _value, _ratio){
        this.error = null
        console.log(`\n ${this.type} Subject: creating a Transfer Request `)
        _to = await _to.getAddress()
        
        const data = ethers.utils.defaultAbiCoder.encode(
            ["uint256"],
            [ethers.utils.parseEther(parseFloat(String(Number(_value) / 100 * _ratio)).toFixed(8))]
        )
        const asset = MOCK_TF_RQ[process.env.REACT_APP_CHAIN].asset
        const transferRequest = new TransferRequest({
            to: _to,
            contractAddress: MOCK_TF_RQ[process.env.REACT_APP_CHAIN].controller.address,
            underwriter: MOCK_TF_RQ[process.env.REACT_APP_CHAIN].trivialUnderwriter,
            module: MOCK_TF_RQ[process.env.REACT_APP_CHAIN].zeroModule,
            asset,
            amount: ethers.utils.parseUnits(String(_value), 8),
            data: String(data)
        })
        this.transferRequest = transferRequest
        return transferRequest
    }

    async load(_transferRequest){
        this.transferRequest = new TransferRequest(_.omit(_transferRequest, ['date']))
        return await this.transferRequest.submitToRenVM()
    }

    async pollTX(_transferRequest){
        this.transferRequest = new TransferRequest(_.omit(_transferRequest, ['date', 'status']))
        this._gatewayAddress = await this.transferRequest.toGatewayAddress()
        this._mint = await this.transferRequest.submitToRenVM()
        this.notify("card_group")

    }

    async sign(signer){
        if (this.error) return
        if (!this.transferRequest || !signer) return this.error = new Error(`${this.type} Subject: ERROR signing: no transfer request or signer`)
        try {
            return this.signature = await this.transferRequest.sign(signer)
        } catch(error){
            console.log("error signing transaction, no wallet connected", error)
        }
    }

    async dry(signer){
        if (this.error) return
        try {
            this._dry = await new TrivialUnderwriterTransferRequest(this.transferRequest).dry(signer.provider, { from : '0x12fBc372dc2f433392CC6caB29CFBcD5082EF494'})
            this._key = await storage.set({...this.transferRequest, date: Date.now()})
        } catch (error) {
            console.log(error)
            this.error = "Loan will fail, double check input values"
            this.timeout = 7000
            this.notify("ERROR")
            return new Error("Loan will fail")
        }
        
    }
    
    async transfer(mock){
        if (this.error) return
        if (mock) return this._gatewayAddress = this._mint.gatewayAddress
        else this._gatewayAddress = await this.transferRequest.toGatewayAddress()
        try {
            await this.zeroUser.publishTransferRequest(this.transferRequest)
            this._mint = await this.transferRequest.submitToRenVM()
            this.notify("CONFIRM")
            this.listener.emit("transaction", this._mint)
        } catch(e) {
            console.log("Error", e)
            if(_.startsWith(this._key, "request:")) window.localStorage.removeItem(`request:${this._key}`)
            else window.localStorage.removeItem(this._key)
            this.error = "Oops, experiencing issues with RenVM"
            this.notify("ERROR")
        }
    }


    async mockSign(signer){
        console.log(`\n ${this.type} Subject: mocking transfer request signature`)
        if (!this.transferRequest || !signer) return this.error = new Error(`\n ${this.type} Subject: error, no transfer request or signer`)
        return this.notify("bridge_group")
    }


}

/**
 * @group bridge_group
 * @events
 *  - transaction_signed
 *      [ pass signed transferRequest to second conversionTool screen ]
 */
export class BridgeObserver extends Observer {
    dispatch
    constructor(_group){
        super(_group)
    }

    async update (monitor) {
        switch (this.group){
            case "CONFIRM":
                let data = monitor.transferRequest
                data.gatewayAddress = monitor._gatewayAddress
                this.dispatch({type: "next", data: { transferRequest: data, back: () => this.dispatch({type: "prev"}) }})
                this.dispatch({type: 'waiting'})
                monitor.listener.on("clear", () => {this.dispatch({type: "prev"})})
                break;
        }
    }
}

export class RefreshObserver extends Observer {
    constructor(_group){
        super(_group)
        this.switch
        this.page
    }
    
    async update (monitor) {
        this.switch()
        this.page(monitor.transferRequest)
    }
}


export class TransactionCardObserver extends Observer {
    constructor(_group){
        super(_group)
        this.append
        this.clear

    }
    async update (monitor) {
        monitor._mint.on('deposit', async (deposit) => {
            const hash = deposit.txHash()
            if (deposit.depositDetails.transaction.confirmations >= 6) return storage.updateTransferRequest(monitor._key, "success")
            await this.append(<TransactionCard btc={monitor._gatewayAddress} confs={deposit}/>)
            (await deposit
                .confirmed())
                .on("target", target => console.log(`0/${target} confirmations`))
                .on("confirmation", (confs, target) => {
                    console.log(`${confs}/${target} confirmations`)
                    if (Number(confs) === 6){
                        storage.updateTransferRequest(monitor._key, "success")
                    }
                })

        })

    }
}

