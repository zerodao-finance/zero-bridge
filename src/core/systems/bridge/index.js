/**
 * @bridge
 * @depricated
 * @contains
 * {
 *  BridgeMonitor
 *  BridgeObserver
 *  RefreshObserver
 *  TransactionCardObserver
 * }
 * 
 * @bridgeEffect
 * @contains
 * {
 *  useBridge
 * }
 * 
 * @bridgeInput
 * @contians
 * {
 *  useBridgeContext
 *  BridgeProvider
 *  
 * }
 * 
 * @transaction
 * @contains
 * {
 *  useTransactionSender
 *  useTransactionListener
 * }
 */
export * from './bridgeEffect'
export * from './bridgeInput'
export * as Hooks from './hooks'
export * as Wrappers from './wrappers'
export * as Context from './context'