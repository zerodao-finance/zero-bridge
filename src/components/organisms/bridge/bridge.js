import { Hooks, Context, useBridgeContext } from '../../../core/systems/bridge'
import { Confirmation, Mechanism } from '../../molecules/bridge'
import { NetworkBoundry } from '../../../core/error'
const { useBridgePage } = Hooks
export const BridgeComponent = () => {
    const state = useBridgePage()
    return (
        <NetworkBoundry>
            <div className="animate-swing-in-top-fwd">
                {
                    state.page == 1 ? <Mechanism _transactionSender={Hooks.useTransactionSender} _context={Context.useBridgeContext} /> : <Confirmation transferRequest={state.data.transferRequest} back={state.data.back} status={state.status}/> 
                }
            </div>
        </NetworkBoundry>
    )
}