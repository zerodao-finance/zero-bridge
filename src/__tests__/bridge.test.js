import renderer from 'react-test-renderer';
import { BridgeComponent } from '../components/organisms/bridge';

describe("Bridge", () => {
    it('renders correctly', () => {
        const bridge = renderer
          .create(<BridgeComponent />)
          .toJSON();
          
        expect(bridge).toMatchSnapshot();
    });
});