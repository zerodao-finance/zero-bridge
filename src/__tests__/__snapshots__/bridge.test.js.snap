// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bridge renders correctly 1`] = `
<h1>
  Something went wrong.
</h1>
`;
